neqo-transport/src/connection/mod.rs

```diff
@@ -2901,6 +2919,43 @@ impl Connection {
             self.decode_ack_delay(ack_delay),
             now,
         );
+
+        let path = self.paths.primary();
+        if path.borrow().is_ecn_enabled() {
+            // RFC 9000, Section 13.4.2.1:
+            //
+            // > An endpoint that receives an ACK frame with ECN counts therefore validates
+            // > the counts before using them. It performs this validation by comparing newly
+            // > received counts against those from the last successfully processed ACK frame.
+            //
+            // RFC 9000 fails to state that this is done *per packet number space*.
+            //
+            // > If an ACK frame newly acknowledges a packet that the endpoint sent with
+            // > either the ECT(0) or ECT(1) codepoint set, ECN validation fails if the
+            // > corresponding ECN counts are not present in the ACK frame.
+            //
+            // We always mark with ECT(0) - if at all - so we only need to check for that.
+            // Also, if we sent a packet with ECT(0) and get only an ACK frame (and not an
+            // ACK-ECN frame), `ecn_counts` will be all zero and the check below will fail,
+            // so no need to explicitly check for the above.
+            //
+            // > ECN validation also fails if the sum of the increase in ECT(0) and ECN-CE counts is
+            // > less than the number of newly acknowledged packets that were originally sent with an
+            // > ECT(0) marking.
+            let newly_acked = acked_packets.len() as u64;
```

Comment:
Please use try_from() rather than `as` where possible.  I think that you can use `?` rather than `.unwrap()` in places that already return `Res`, but `.unwrap()` is totally fine for `u64` <-> `usize` conversions..

Reply:
Thanks, will fix. `as` just seemed so much more readable, and I didn't know about the downsides.